src/minion-ffi/tests/runner.c

```
@@ -0,0 +1,215 @@
+    verify_ok(
+        minion_cp_spawn(bk,
+                        (struct Minion_ChildProcessOptions) {
+                            .image_path = "/me",
+                            .argv = (const char*[]) {test_name},
+                            .envp = (struct Minion_EnvItem[1]) {ENV_ITEM_FIN},
+                            .stdio = {devnull_fd, dup(1), dup(1)},
+                            .dominion = sandbox,
+                            .workdir = "/"},
+                        &proc));
+    Minion_WaitOutcome outcome;
+    verify_ok(
+        minion_cp_wait(proc, &(const struct Minion_TimeSpec) {100, 0}, &outcome));
+    if (outcome == WAIT_OUTCOME_TIMEOUT) {
+        verify_ok(minion_cp_kill(proc));
+        assert_write(1, "ILE\n", 4);
+    } else if (outcome == WAIT_OUTCOME_ALREADY_FINISHED)
+        assert_write(1, "Already finished, WTF?\n", 23);
+    else {
+        int64_t exitcode = 57179444;
+
+        verify_ok(minion_cp_exitcode(proc, &exitcode, NULL));
+        char data[64];
+        assert_write(1, data,
+                     sprintf(data, "exit code %lld\n", (long long) exitcode));
+    }
+    verify_ok(minion_cp_free(proc));
+    verify_ok(minion_dominion_free(sandbox));
+    verify_ok(minion_backend_free(bk));
+    exit(0);
+}
+
+#include "tests.h"
+
+int test_main(int argc, char const* const* argv) {
+    if (argc != 2)
+        abort();
+    for (int i = 0; tests[i].name; i++)
+        if (!strcmp(tests[i].name, argv[1])) {
+            tests[i].func();
+            die("program has not exited after running test");
+        }
+    fprintf(stderr, "test %s not found", argv[1]);
+    return 179;
+}
+
+size_t read_all(int fd, const char** ans_p)
+{
+    char* ans = NULL;
+    size_t cap = 0;
+    size_t sz = 0;
+    for(;;)
+    {
+        if(sz == cap)
+        {
+            cap = 2 * cap + 1;
+            ans = realloc(ans, cap);
+        }
+        ssize_t chunk_sz = read(fd, ans + sz, cap - sz);
+        if(chunk_sz <= 0)
```

Why error can be ignored?